File: docs/listings/marketplace/client-sdk.ts
/**
 * Script Master Labs — mcp-x402 TypeScript client
 * Generated 2026-07-29 from live OpenAPI (77 paths)
 * SAM UEI G24VZA4RLMK3 · CAGE 21U51 · SDVOSB
 * Base: https://mcp-x402.onrender.com
 * x402 payTo: 0x4e14B249D9A4c9c9352D780eCEB508A8eB7a7700 (eip155:8453 USDC)
 */
export type SmlClientOptions = {
  baseUrl?: string;
  apiKey?: string;
  /** raw x402 payment header value if you already settled */
  paymentHeader?: string;
  fetchImpl?: typeof fetch;
};

export class SmlX402Client {
  readonly baseUrl: string;
  private apiKey?: string;
  private paymentHeader?: string;
  private fetchImpl: typeof fetch;

  constructor(opts: SmlClientOptions = {}) {
    this.baseUrl = (opts.baseUrl || "https://mcp-x402.onrender.com").replace(/\/$/, "");
    this.apiKey = opts.apiKey;
    this.paymentHeader = opts.paymentHeader;
    this.fetchImpl = opts.fetchImpl || fetch;
  }

  async request<T = unknown>(method: string, path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers || {});
    headers.set("Accept", "application/json");
    headers.set("User-Agent", "sml-x402-sdk-ts/1.0");
    if (this.apiKey) headers.set("Authorization", `Bearer ${this.apiKey}`);
    if (this.paymentHeader) {
      headers.set("X-PAYMENT", this.paymentHeader);
      headers.set("PAYMENT-SIGNATURE", this.paymentHeader);
    }
    const res = await this.fetchImpl(`${this.baseUrl}${path}`, { ...init, method, headers });
    if (res.status === 402) {
      const body = await res.json().catch(() => ({}));
      const err = new Error("Payment Required (x402)") as Error & { status: number; challenge: unknown };
      err.status = 402;
      err.challenge = body;
      throw err;
    }
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`${method} ${path} → ${res.status} ${text.slice(0, 300)}`);
    }
    const ct = res.headers.get("content-type") || "";
    if (ct.includes("application/json")) return (await res.json()) as T;
    return (await res.text()) as unknown as T;
  }

  getOpenAPI() { return this.request<Record<string, unknown>>("GET", "/openapi.json"); }
  getCatalog() { return this.request<Record<string, unknown>>("GET", "/.well-known/x402"); }
  health() { return this.request<Record<string, unknown>>("GET", "/health"); }
  stats() { return this.request<Record<string, unknown>>("GET", "/x402/stats"); }

  /** Generic paid snack helper — path must start with /x402/ */
  snack<T = unknown>(path: string, query?: Record<string, string>) {
    const qs = query ? "?" + new URLSearchParams(query).toString() : "";
    const p = path.startsWith("/") ? path : `/${path}`;
    return this.request<T>("GET", `${p}${qs}`);
  }
}

export default SmlX402Client;

File: downloads/api-marketplace/client_sdk.py
#!/usr/bin/env python3
"""Script Master Labs — mcp-x402 Python client
Generated 2026-07-29 from live OpenAPI (77 paths)
SAM UEI G24VZA4RLMK3 · CAGE 21U51 · SDVOSB
Base: https://mcp-x402.onrender.com
x402 payTo: 0x72330994f379a71542e7bd5a4cf99a9d9743f4aa (eip155:8453 USDC)
"""
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, Optional

DEFAULT_BASE = "https://mcp-x402.onrender.com"
PAY_TO = "0x72330994f379a71542e7bd5a4cf99a9d9743f4aa"
SAM_UEI = "G24VZA4RLMK3"
CAGE = "21U51"


class PaymentRequired(Exception):
    def __init__(self, challenge: Any):
        super().__init__("Payment Required (x402)")
        self.challenge = challenge
        self.status = 402


class SmlX402Client:
    def __init__(
        self,
        base_url: str = DEFAULT_BASE,
        api_key: Optional[str] = None,
        payment_header: Optional[str] = None,
        timeout: float = 30.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.payment_header = payment_header
        self.timeout = timeout

    def request(self, method: str, path: str, query: Optional[Dict[str, str]] = None) -> Any:
        if not path.startswith("/"):
            path = "/" + path
        url = self.base_url + path
        if query:
            url += "?" + urllib.parse.urlencode(query)
        headers = {
            "Accept": "application/json",
            "User-Agent": "sml-x402-sdk-py/1.0",
        }
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        if self.payment_header:
            headers["X-PAYMENT"] = self.payment_header
            headers["PAYMENT-SIGNATURE"] = self.payment_header
        req = urllib.request.Request(url, headers=headers, method=method.upper())
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read()
                ct = resp.headers.get("Content-Type", "")
                if "application/json" in ct:
                    return json.loads(raw.decode() or "null")
                return raw.decode()
        except urllib.error.HTTPError as e:
            body = e.read()
            if e.code == 402:
                try:
                    challenge = json.loads(body.decode() or "{}")
                except Exception:
                    challenge = body.decode(errors="ignore")
                raise PaymentRequired(challenge) from e
            raise RuntimeError(f"{method} {path} → {e.code} {body[:300]!r}") from e

    def get_openapi(self) -> dict:
        return self.request("GET", "/openapi.json")

    def get_catalog(self) -> dict:
        return self.request("GET", "/.well-known/x402")

    def health(self) -> dict:
        return self.request("GET", "/health")

    def stats(self) -> dict:
        return self.request("GET", "/x402/stats")

    def snack(self, path: str, query: Optional[Dict[str, str]] = None) -> Any:
        """GET a paid snack path (e.g. /x402/gas-tracker)."""
        return self.request("GET", path, query=query)


if __name__ == "__main__":
    c = SmlX402Client()
    print(json.dumps(c.health(), indent=2))
    oa = c.get_openapi()
    print("paths", len(oa.get("paths") or {}), "uei", SAM_UEI, "cage", CAGE)
